### ISSUE TITLE
"undefined variable" error when updating from 2.8.2 to 2.9.4

### ISSUE THREAD

**Issue by zaskan** [ansible/ansible#66977]
##### SUMMARY

Appears that some variables are not processed properly when update to 2.9 version. In this case I think that the problem is related with how Ansible is processing the variable:

`eap_ver_full: "{{ eap_ver_major }}.{{ eap_ver_minor }}.{{ eap_ver_patch }}"`

We tried to find out information about changes in the release notes but we couldn't find anything.

We really appreciate your help.

Thanks so much.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

Ansible engine
##### ANSIBLE VERSION

```paste below
2.8.2 and 2.9.4
```
##### STEPS TO REPRODUCE

**inventory.yml**
~~~
all:
  vars:
    eap:
      '7':
        ver: 7.2.6
  hosts:
    localhost:
~~~
**main.yml**
~~~
---
- name: TEST Ansible 2.9 bug
  hosts: localhost
  connection: local
  gather_facts: false

  vars:
#    play_eap:
#      ver: 7.2.6

    play_eap_name: '7'
    play_eap_list: "{{ play_prj.eap }}"
    play_eap: "{{ play_eap_list[play_eap_name] }}"
    play_prj: "{{ hostvars[inventory_hostname] }}"

  tasks:
#  - debug:
#      var: play_eap.ver

#  - meta: end_play

  - name: SETUP | EAP product
    include_role:
      name: "eap"
    vars:
      eap_args_major: "{{ play_eap.ver.split('.')[0] }}"
      eap_args_minor: "{{ play_eap.ver.split('.')[1] }}"
      eap_args_patch: "{{ play_eap.ver.split('.')[2] }}"
      eap_args_bin_lv: "{{ play_eap.bin_lv | default(play_eap_list.bin_lv) | default(omit) }}"
      eap_args_bin_path: "{{ play_eap.bin_path | default(play_eap_list.bin_path) | default(omit) }}"

~~~
**roles/eap/varsmain.yml**
~~~
eap_ver_full: "{{ eap_ver_major }}.{{ eap_ver_minor }}.{{ eap_ver_patch }}"
eap_ver_major: "{{ eap_args_major }}"
eap_ver_minor: "{{ eap_args_minor }}"
eap_ver_patch: "{{ eap_args_patch }}"
eap_fact_ver_full: 7.2.6
~~~
**roles/eap/varseap_exists.yml**
~~~
eap_cnd_absent: false
eap_fact_ver_equal: "{{ eap_ver_full is version(eap_fact_ver_full, '==') }}"
eap_fact_ver_major: "{{ eap_fact_ver_full.split('.')[0] }}"
eap_fact_ver_minor: "{{ eap_fact_ver_full.split('.')[1] }}"
eap_fact_ver_patch: "{{ eap_fact_ver_full.split('.')[2] }}"
eap_fact_ver_pkg: "{{ eap_fact_ver_major }}.{{ eap_fact_ver_minor }}"
~~~
**roles/eap/tasks/main.yml**
~~~
- name: get facts
  include_tasks:
    file: get_facts.yml
    apply:
      tags:
      - always

- name: initialize
  include_tasks:
    file: init.yml
    apply:
      tags:
      - always
~~~
**roles/eap/tasks/get_facts.yml**
~~~
- block:
  - name: load variables for existing eap product
    include_vars:
      file: eap_exists.yml
  when: true
~~~
**roles/eap/tasks/init.yml**
~~~
- name: debug args
  debug:
    msg: |
         eap_args_name="{{ eap_args_name | default('<UNDEFINED>') }}"
         eap_args_major="{{ eap_args_major | default('<UNDEFINED>') }}"
         eap_args_minor="{{ eap_args_minor | default('<UNDEFINED>') }}"
         eap_args_patch="{{ eap_args_patch | default('<UNDEFINED>') }}"
         eap_args_bin_lv="{{ eap_args_bin_lv | default('<UNDEFINED>') }}"
         eap_args_bin_path="{{ eap_args_bin_path | default('<UNDEFINED>') }}"
         eap_args_bin_size="{{ eap_args_bin_size | default('<UNDEFINED>') }}"
         eap_args_bin_vg="{{ eap_args_bin_vg | default('<UNDEFINED>') }}"
         eap_args_log_lv="{{ eap_args_log_lv | default('<UNDEFINED>') }}"
         eap_args_log_path="{{ eap_args_log_path | default('<UNDEFINED>') }}"
         eap_args_log_size="{{ eap_args_log_size | default('<UNDEFINED>') }}"
         eap_args_log_vg="{{ eap_args_log_vg | default('<UNDEFINED>') }}"
         eap_args_user_id="{{ eap_args_user_id | default('<UNDEFINED>') }}"
         eap_args_user_name="{{ eap_args_user_name | default('<UNDEFINED>') }}"
         eap_args_group_id="{{ eap_args_group_id | default('<UNDEFINED>') }}"
         eap_args_group_name="{{ eap_args_group_name | default('<UNDEFINED>') }}"
         eap_args_mod_force="{{ eap_args_mod_force | default('<UNDEFINED>') }}"
         eap_args_mod_list="{{ eap_args_mod_list | default('<UNDEFINED>') }}"
         eap_args_state="{{ eap_args_state | default('<UNDEFINED>') }}"
         eap_args_force="{{ eap_args_force | default('<UNDEFINED>') }}"
  tags:
  - never
  - debug

- name: debug initial state
  debug:
    msg: |
         eap_cnd_absent="{{ eap_cnd_absent | default('<UNDEFINED>') }}"
         eap_cnd_install="{{ eap_cnd_install | default('<UNDEFINED>') }}"
         eap_cnd_patch="{{ eap_cnd_patch | default('<UNDEFINED>') }}"
         eap_cnd_replace="{{ eap_cnd_replace | default('<UNDEFINED>') }}"
         eap_cnd_uninstall="{{ eap_cnd_uninstall | default('<UNDEFINED>') }}"

         eap_fact_ver_equal="{{ eap_fact_ver_equal | default('<UNDEFINED>') }}"
         eap_fact_ver_full="{{ eap_fact_ver_full | default('<UNDEFINED>') }}"
         eap_fact_ver_major="{{ eap_fact_ver_major | default('<UNDEFINED>') }}"
         eap_fact_ver_minor="{{ eap_fact_ver_minor | default('<UNDEFINED>') }}"
         eap_fact_ver_patch="{{ eap_fact_ver_patch | default('<UNDEFINED>') }}"
         eap_fact_ver_pkg="{{ eap_fact_ver_pkg | default('<UNDEFINED>') }}"

         eap_name="{{ eap_args_name | default('<UNDEFINED>') }}"
         eap_major="{{ eap_args_major | default('<UNDEFINED>') }}"
         eap_minor="{{ eap_args_minor | default('<UNDEFINED>') }}"
         eap_patch="{{ eap_args_patch | default('<UNDEFINED>') }}"
         eap_bin_lv="{{ eap_args_bin_lv | default('<UNDEFINED>') }}"
         eap_bin_path="{{ eap_args_bin_path | default('<UNDEFINED>') }}"
         eap_bin_size="{{ eap_args_bin_size | default('<UNDEFINED>') }}"
         eap_bin_vg="{{ eap_args_bin_vg | default('<UNDEFINED>') }}"
         eap_log_lv="{{ eap_args_log_lv | default('<UNDEFINED>') }}"
         eap_log_path="{{ eap_args_log_path | default('<UNDEFINED>') }}"
         eap_log_size="{{ eap_args_log_size | default('<UNDEFINED>') }}"
         eap_log_vg="{{ eap_args_log_vg | default('<UNDEFINED>') }}"
         eap_user_id="{{ eap_args_user_id | default('<UNDEFINED>') }}"
         eap_user_name="{{ eap_args_user_name | default('<UNDEFINED>') }}"
         eap_group_id="{{ eap_args_group_id | default('<UNDEFINED>') }}"
         eap_group_name="{{ eap_args_group_name | default('<UNDEFINED>') }}"
         eap_mod_force="{{ eap_args_mod_force | default('<UNDEFINED>') }}"
         eap_mod_list="{{ eap_args_mod_list | default('<UNDEFINED>') }}"
         eap_state="{{ eap_args_state | default('<UNDEFINED>') }}"
         eap_force="{{ eap_args_force | default('<UNDEFINED>') }}"
  tags:
  - never
  - debug
~~~

##### EXPECTED RESULTS

Variables should be stored as in 2.8.X

##### ACTUAL RESULTS

`An unhandled exception occurred while templating '{{ play_eap.bin_vg | default(play_eap_list.bin_vg) | default(omit) }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: An unhandled exception occurred while templating '{{ play_eap_list[play_eap_name] }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: An unhandled exception occurred while templating '{{ play_prj.eap }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: An unhandled exception occurred while templating '{{ hostvars[inventory_hostname] }}'. Error was a <class 'ansible.errors.AnsibleFilterError'>, original message: Version comparison: 'eap_ver_full' is undefined`

```paste below

```

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by stmarier** [comment]
I can confirm this is happening to me as well upgrading from 2.8.0 to 2.9.4

In one task we discover stack outputs from cloudformation using `cloudformation_info`.  We can use a debugger/debug module to get the data we need through a normal lookup chain (ex `register['cloudformation'][stack_name][etc]`.

The next task is to push another cloudformation template using `lookup('template')` and the lookup chain fails due to `register` not having a value for `['cloudformation']`.

This is a pretty big regression! Let me know if I can help provide examples or test any fixes.

**Comment by mkrizek** [comment]
I have `git bisect`'ed this to https://github.com/ansible/ansible/pull/64282. 

We seem to avoid failing in `HostVarsVars.repr()` on an undefined variable, calling `Templar.template()` with `fail_on_undefined=False`:
https://github.com/ansible/ansible/blob/0ee28c14c21cdddbf03842afb350ef3a1a83f899/lib/ansible/vars/hostvars.py#L152-L154

The actual error in this issue is that the undefined error is wrapped in `AnsibleFilterError` which is why it fails:
https://github.com/ansible/ansible/blob/0ee28c14c21cdddbf03842afb350ef3a1a83f899/lib/ansible/template/__init__.py#L663-L667

cc @s-hertel

**Comment by zaskan** [comment]
Hi all. Is there any date expected for fixing it? Thanks!

**Comment by stefanoabbate** [comment]
Are there any workarounds for the reproducer example?

**Comment by mkrizek** [comment]
I may have took it too far but here's simpler reproducer:

playbook.yml
```yaml
- hosts: localhost
  gather_facts: no
  vars:
    play_prj: "{{ hostvars[inventory_hostname] }}"
    eap_args_major: "{{ play_prj.eap['7'].ver.split('.')[0] }}"
  roles:
    - eap
```

inventory.yml
```yaml
all:
  vars:
    eap:
      '7':
        ver: 7.2.6
  hosts:
    localhost:
```

roles/eap/tasks/main.yml
```yaml
- include_vars:
    file: eap_exists.yml

- debug:
    msg: "{{ eap_args_major  }}"
```

roles/eap/vars/main.yml
```yaml
eap_ver_full: "{{ eap_args_major }}"
eap_fact_ver_full: 7
```

roles/eap/vars/eap_exists.yml
```yaml
eap_fact_ver_equal: "{{ eap_ver_full is version(7, '==') }}"
```

```
$ ansible-playbook -i inventory.yml playbook.yml
PLAY [localhost] ****************************************************************************************************************************

TASK [eap : include_vars] *******************************************************************************************************************
ok: [localhost]

TASK [eap : debug] **************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while templating '{{ play_prj.eap['7'].ver.split('.')[0] }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: An unhandled exception occurred while templating '{{ hostvars[inventory_hostname] }}'. Error was a <class 'ansible.errors.AnsibleFilterError'>, original message: Version comparison: 'eap_ver_full' is undefined"}

PLAY RECAP **********************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

The workaround for this is not to template the whole `hostvars[inventory_name]` and just access the attributes directly, so changing `playbook.yml` to:
```yaml
- hosts: localhost
  gather_facts: no
  vars:
    eap_args_major: "{{ hostvars[inventory_hostname].eap['7'].ver.split('.')[0] }}"
  roles:
    - eap
```

results in:

```
$ ansible-playbook -i inventory.yml playbook.yml
PLAY [localhost] ****************************************************************************************************************************

TASK [eap : include_vars] *******************************************************************************************************************
ok: [localhost]

TASK [eap : debug] **************************************************************************************************************************
ok: [localhost] => {
    "msg": "7"
}

PLAY RECAP **********************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

**Comment by s-hertel** [comment]
It seems like this is specific to a number of Ansible's filters and the version/version_compare test because we don't always preserve UndefinedError. It looks like the jinja2 builtins work. One solution could be fixing the individual filters
```diff
diff --git a/lib/ansible/plugins/test/core.py b/lib/ansible/plugins/test/core.py
index 66aff2a14b..6390dc6612 100644
--- a/lib/ansible/plugins/test/core.py
+++ b/lib/ansible/plugins/test/core.py
@@ -158,9 +158,12 @@ def version_compare(value, version, operator='eq', strict=False):
     else:
         raise errors.AnsibleFilterError('Invalid operator type')
 
+    from jinja2.exceptions import UndefinedError
     try:
         method = getattr(py_operator, operator)
         return method(Version(str(value)), Version(str(version)))
+    except UndefinedError as e:
+        raise errors.AnsibleUndefinedVariable('Version comparison: %s' % e)
     except Exception as e:
         raise errors.AnsibleFilterError('Version comparison: %s' % e)
```
I'm not sure if this could mask too much, but it might be a better approach:
```diff
diff --git a/lib/ansible/vars/hostvars.py b/lib/ansible/vars/hostvars.py
index 1d3c5ef943..996d524c12 100644
--- a/lib/ansible/vars/hostvars.py
+++ b/lib/ansible/vars/hostvars.py
@@ -141,4 +141,6 @@ class HostVarsVars(Mapping):
 
     def __repr__(self):
         templar = Templar(variables=self._vars, loader=self._loader)
+        templar._fail_on_filter_errors = False
         return repr(templar.template(self._vars, fail_on_undefined=False, static_vars=STATIC_VARS))
```

**Comment by stefanoabbate** [comment]
@mkrizek sorry for getting back to you so late. Thank you for your solution. I tried it and it was working... however I found a way to work around the issue by changing a little part of the playbook.

Basically, I removed the file eap_exists.yml and set roles/eap/tasks/main.yml as follows:

```
- set_fact:
    eap_fact_ver_equal: "{{ eap_ver_full is version(eap_fact_ver_full, '==') }}"

- debug:
    msg: "{{ eap_args_major  }}"
```

I used set_fact instead of dynamically include vars. In the original code the "include vars" was under a "when condition" so the same can be done with set_fact.
